npm/postinstall.ts

```diff
@@ -0,0 +1,211 @@
+import path from "path";
+import request from "request";
+import tar from "tar";
+import unzipper from "unzipper";
+import { URL } from "url";
+
+/**
+ * A map of Node's `process.arch` to Golang's `$GOARCH` equivalent.
+ */
+const archMap: Partial<Record<typeof process.arch, string>> = {
+  arm64: "arm64",
+  x64: "x86_64",
+};
+
+/**
+ * Known extensions that we package our binaries with.
+ *
+ * In an enum so that we can be type-safe with Record<> and ensure we're always
+ * handling any extensions we add here.
+ */
+enum KnownExtension {
+  Tar = ".tar.gz",
+  Zip = ".zip",
+}
+
+/**
+ * A map of Node's `process.platform` to Golang's `$GOOS` equivalent and the
+ * extension of the packaged binary we provide.
+ */
+const platformMap: Partial<
+  Record<
+    typeof process.platform,
+    { platform: string; extension: KnownExtension }
+  >
+> = {
+  linux: { platform: "linux", extension: KnownExtension.Tar },
+  darwin: { platform: "darwin", extension: KnownExtension.Tar },
+  win32: { platform: "windows", extension: KnownExtension.Zip },
+};
+
+/**
+ * Fetches the supposed URL of the binary on GitHub for this version.
+ */
+async function getBinaryUrl(): Promise<URL> {
+  const version = process.env.npm_package_version?.trim();
+
+  if (!version) {
+    throw new Error("Could not find package version to install binary");
+  }
+
+  const { arch, platform } = getArchPlatform();
+
+  const targetUrl = new URL(
+    `https://cli.inngest.com/artifact/v${version}/inngest_${version}_${platform.platform}_${arch}${platform.extension}`
+  );
+
+  return targetUrl;
+}
+
+/**
+ * Return the platform and architecture included in the supposed binary name
+ * from `$GOOS` and `$GOARCH`.
+ *
+ * @throws when the arch or platform are not supported by this script.
+ */
+function getArchPlatform() {
+  const arch = archMap[process.arch];
+  const platform = platformMap[process.platform];
+
+  if (!arch) {
+    throw new Error(`Unsupported architecture: ${process.arch}`);
+  }
+
+  if (!platform) {
+    throw new Error(`Unsupported platform: ${process.platform}`);
+  }
+
+  return { arch, platform };
+}
+
+/**
+ * Attempts to fetch a binary given a `url` and returns the response so long as
+ * it is a successful `200 OK`.
+ *
+ * @throws when the response is not `200 OK`.
+ */
+function downloadBinary(
+  /**
+   * The URL to download the binary from.
+   */
+  url: URL
+): Promise<request.Request> {
+  /**
+   * Using `new Promise()` here so that we can use listeners when dealing with
+   * the request in order to ensure it's valid without using the body.
+   */
+  return new Promise((resolve, reject) => {
+    const req = request({ uri: url.href });
```

Comment:
Aha - yes indeed. So there are a few alternatives, namely the core `https` lib, `fetch` (via `node-fetch` as we're on `v16`), or `axios`.

`https` doesn't follow redirects, even with its lovely `https.request()` function. GitHub seems to redirect at least once on the way to the download; I'd rather skip that code if I can!

I can't remember why I vetoed `fetch` and `axios`; it may have been that I discounted them based on earlier code, and they'd be a reasonable fit now.

I'll have a check to see if we can use those instead.

Reply:
`node-fetch` looks like it's working well. Will move to that instead. 👍